backend/corpora/dataset_processing/remaster_cxg.py

```diff
@@ -0,0 +1,250 @@
+    :return:
+    """
+
+    # Careful with old datasets - the ones that do not have a UUID
+    # cxg_filename = f"{dataset_id}.cxg"
+
+    # Download the cxg from the bucket
+    bucket_prefix = get_bucket_prefix(dataset_id)
+    object_key = f"{bucket_prefix}.cxg"
+    path = f"s3://{cellxgene_bucket}/{object_key}/X"
+    local_path = "/cxg"
+    logger.info(f"Downloading {path} to {local_path}/X_old")
+
+    download_command = ["aws", "s3", "sync", path, f"{local_path}/X_old"]
+    # Let errors fail the pipeline
+    subprocess.run(download_command, check=True)
+
+    kind = "auto"
+    obs_extent = 256
+    var_extent = 2048
+    cell_order = "row"
+    tile_order = "col"
+    capacity = 128000
+    compression = 22
+    target_array = "X_new"
+    source_array = "X_old"
+    sparse_threshold = 25.0
+
+    compute(
+        local_path,
+        obs_extent,
+        var_extent,
+        source_array,
+        target_array,
+        kind,
+        sparse_threshold,
+        compression,
+        cell_order,
+        tile_order,
+        capacity,
+    )
+
+    upload_command = ["aws", "s3", "sync", f"{local_path}/X_new", path]
+    subprocess.run(upload_command, check=True)
+
+def compute(
+    cxg,
+    obs_extent,
+    var_extent,
+    source_array,
+    target_array,
+    kind,
+    sparse_threshold,
+    compression,
+    cell_order,
+    tile_order,
+    capacity,
+):
+    X_extent = [obs_extent, var_extent]
+
+    ctx = create_fast_ctx(
+        {
+            "py.init_buffer_bytes": 32 * 1024**3,
+            "sm.tile_cache_size": 2 * 1024**3,
+        }
+    )
+
+    if kind == "auto":
+        """
+        Determine target encoding (sparse or dense) from the data.
+        The "magic" sparsity threshold should in sync with
+        backend/corpora/dataset_processing/process.py::make_cxg()
+        """
+        if not 0 < sparse_threshold <= 100:
+            logger.info("Sparse threhold must be in range (0, 100]")
+            return
+        kind = choose_X_encoding(cxg, source_array, sparse_threshold, ctx)
+
+    out_sparse = kind == "sparse"
+
+    st = time.time()
+    with tiledb.scope_ctx(ctx):
+        with tiledb.open(f"{cxg}/{source_array}", "r") as old_X:
+            create_new_X(
+                cxg,
+                target_array,
+                kind,
+                compression,
+                X_extent,
+                old_X.schema,
+                cell_order,
+                tile_order,
+                capacity,
+            )
+            logger.info("created, starting to read...")
+            in_sparse = old_X.schema.sparse
+            with tiledb.open(f"{cxg}/{target_array}", "w") as new_X:
+                logger.info(f"sparsity analysis: {in_sparse} {out_sparse} {old_X.schema.sparse} {new_X.schema.sparse}")
+                if in_sparse and out_sparse:
+                    logger.info("sparse->sparse")
+                    i, chunk = 0, 200_000
+                    while i < old_X.shape[0]:
```

Comment:
side note for future reference:  TileDB has incremental queries which obviate the need to do this kind of iterative & chunked reading.  No reason to change this code, but I wouldn't use this style for future/new code.  See the `return_incomplete` option on the tiledb.Array.query() function for more info.